oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/DefaultOAuth2AuthenticatedPrincipal.java

```diff
@@ -81,6 +81,11 @@ public Collection<? extends GrantedAuthority> getAuthorities() {
 		return this.authorities;
 	}
 
+	@Override
```

Comment:
What if we instead added a constructor that took the `String name` instead of the attribute key? Then, we could simply deprecate the `nameAttributeKey` constructor and have it call the new constructor? In that case, this method and the interface method would go away.

Reply:
At that point, we'd also want to update production files to use the new constructor (that is, tests can stay the same).

Reply:
I like your suggestion with a new constructor. It wouldn't affect downstream users of the interface and aligns well with the getters since there would be one constructor parameter for each.

The only issue I see is that the existing `String nameAttributeKey` and new `String name` constructors would share the same unless the order was changed to have the `name` in front:

```java
public DefaultOAuth2User(String name, Collection<? extends GrantedAuthority> authorities, Map<String, Object> attributes)
```

Reply:
Good point. Typically authenticated principals in Spring Security follow this order:

```
(principal, details, authorities)
```

So this is also an opportunity to bring this constructor into alignment with the rest. It should be:

```
(String name, Map<String, Object> attributes, Collection<? extends GrantedAuthority> authorities)
```

In that case, will you also please add a `@JsonCreator` in `DefaultOAuth2UserMixin` that points to the new constructor? Note that the old `@JsonCreator` and constructor should remain in place.

Reply:
I added the new `String name` constructors to `DefaultOAuth2User` and `DefaultOidcUser` and accompanying `@JsonCreator` in `DefaultOAuth2UserMixin` and `DefaultOidcUserMixin`. `DefaultOidcUser` has quite a few constructors and I aligned the parameter order in all of them, deprecating the old ones.

All production usages use the new constructors (except for one in `OidcUserRequestUtils.getUser` which would require `OidcUserAuthority.collectClaims` to be accessible, so this is still a to-do) while the tests stay the same.

One last open topic would be unit tests for the new constructors, but I figured I'd wait for your feedback before approaching them.